pkg/distributor/push.go

```diff
@@ -306,44 +331,50 @@ func calculateRetryAfter(retryAttemptHeader string, baseSeconds int, maxBackoffE
 // toGRPCHTTPStatus converts the given error into an appropriate GRPC and HTTP status corresponding
 // to that error, if the error is one of the errors from this package. Otherwise, codes.Internal and
 // http.StatusInternalServerError is returned.
-func toGRPCHTTPStatus(ctx context.Context, pushErr error, limits *validation.Overrides) (codes.Code, int) {
-	if errors.Is(pushErr, context.DeadlineExceeded) {
-		return codes.Internal, http.StatusInternalServerError
-	}
-
-	var distributorErr Error
-	if errors.As(pushErr, &distributorErr) {
-		switch distributorErr.Cause() {
-		case mimirpb.BAD_DATA:
-			return codes.InvalidArgument, http.StatusBadRequest
-		case mimirpb.INGESTION_RATE_LIMITED, mimirpb.REQUEST_RATE_LIMITED:
-			serviceOverloadErrorEnabled := false
-			userID, err := tenant.TenantID(ctx)
-			if err == nil {
-				serviceOverloadErrorEnabled = limits.ServiceOverloadStatusCodeOnRateLimitEnabled(userID)
-			}
-			// Return a 429 or a 529 here depending on configuration to tell the client it is going too fast.
-			// Client may discard the data or slow down and re-send.
-			// Prometheus v2.26 added a remote-write option 'retry_on_http_429'.
-			if serviceOverloadErrorEnabled {
-				return codes.ResourceExhausted, StatusServiceOverloaded
+// convertHTTPCodeFunc is utilized by the OTLP endpoint.
+// According to the OTLP specifications (https://opentelemetry.io/docs/specs/otlp/#failures-1), unlike Prometheus, the OTLP client only retries on HTTP status codes 429, 502, 503, and 504.
+// Use convertHTTPCodeFunc to map non-retryable HTTP error codes to retryable ones.
+func toGRPCHTTPStatus(ctx context.Context, pushErr error, limits *validation.Overrides, convertHTTPCodeFunc func(httpCode int) int) (codes.Code, int) {
```

Comment:
Would it make sense to split this function into Prometheus and OTLP version?

Reply:
trying to avoid dublicated the code here, would we prefer dublicate and keep it more straight forward?

Reply:
Since the mapping to HTTP status codes is different between Prometheus remote-write and OTLP, I think it makes sense to have two versions and simplify the code.

Reply:
Do we _need_ to use codes different to what OTLP accepts?